<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Statistical Genetics: Analyses with R</title>
  <meta name="description" content="Description to come…">
  <meta name="generator" content="bookdown 0.3.9 and GitBook 2.6.7">

  <meta property="og:title" content="Statistical Genetics: Analyses with R" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Description to come…" />
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Statistical Genetics: Analyses with R" />
  
  <meta name="twitter:description" content="Description to come…" />
  

<meta name="author" content="Dr. Shirin Glander">


<meta name="date" content="2017-04-03">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="population-genetics.html">
<link rel="next" href="genetics-of-complex-diseases.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #f8f8f8; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
pre, code { background-color: #f8f8f8; }
code > span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code > span.dt { color: #204a87; } /* DataType */
code > span.dv { color: #0000cf; } /* DecVal */
code > span.bn { color: #0000cf; } /* BaseN */
code > span.fl { color: #0000cf; } /* Float */
code > span.ch { color: #4e9a06; } /* Char */
code > span.st { color: #4e9a06; } /* String */
code > span.co { color: #8f5902; font-style: italic; } /* Comment */
code > span.ot { color: #8f5902; } /* Other */
code > span.al { color: #ef2929; } /* Alert */
code > span.fu { color: #000000; } /* Function */
code > span.er { color: #a40000; font-weight: bold; } /* Error */
code > span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #000000; } /* Constant */
code > span.sc { color: #000000; } /* SpecialChar */
code > span.vs { color: #4e9a06; } /* VerbatimString */
code > span.ss { color: #4e9a06; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #000000; } /* Variable */
code > span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code > span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code > span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code > span.ex { } /* Extension */
code > span.at { color: #c4a000; } /* Attribute */
code > span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code > span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a></li>
<li class="chapter" data-level="" data-path="about-the-author.html"><a href="about-the-author.html"><i class="fa fa-check"></i>About the Author</a></li>
<li class="chapter" data-level="1" data-path="introduction-to-statistical-genetics.html"><a href="introduction-to-statistical-genetics.html"><i class="fa fa-check"></i><b>1</b> Introduction to Statistical Genetics</a></li>
<li class="chapter" data-level="2" data-path="evolutionary-genetics.html"><a href="evolutionary-genetics.html"><i class="fa fa-check"></i><b>2</b> Evolutionary Genetics</a></li>
<li class="chapter" data-level="3" data-path="population-genetics.html"><a href="population-genetics.html"><i class="fa fa-check"></i><b>3</b> Population Genetics</a></li>
<li class="chapter" data-level="4" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html"><i class="fa fa-check"></i><b>4</b> Quantitative Genetics</a><ul>
<li class="chapter" data-level="4.1" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html#qtl-analysis"><i class="fa fa-check"></i><b>4.1</b> QTL analysis</a><ul>
<li class="chapter" data-level="4.1.1" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html#recombinant-inbred-lines-rils"><i class="fa fa-check"></i><b>4.1.1</b> Recombinant Inbred Lines (RILs)</a></li>
<li class="chapter" data-level="4.1.2" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html#qtl-analysis-in-r"><i class="fa fa-check"></i><b>4.1.2</b> QTL analysis in R</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html#gene-x-environment-interactions"><i class="fa fa-check"></i><b>4.2</b> Gene x Environment interactions</a></li>
<li class="chapter" data-level="4.3" data-path="quantitative-genetics.html"><a href="quantitative-genetics.html#variance-and-heritability"><i class="fa fa-check"></i><b>4.3</b> Variance and heritability</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="genetics-of-complex-diseases.html"><a href="genetics-of-complex-diseases.html"><i class="fa fa-check"></i><b>5</b> Genetics of complex diseases</a></li>
<li class="chapter" data-level="6" data-path="genetic-epidemiology.html"><a href="genetic-epidemiology.html"><i class="fa fa-check"></i><b>6</b> Genetic Epidemiology</a></li>
<li class="chapter" data-level="7" data-path="placeholder.html"><a href="placeholder.html"><i class="fa fa-check"></i><b>7</b> Placeholder</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Statistical Genetics: Analyses with R</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="quantitative-genetics" class="section level1">
<h1><span class="header-section-number">4</span> Quantitative Genetics</h1>
<div id="qtl-analysis" class="section level2">
<h2><span class="header-section-number">4.1</span> QTL analysis</h2>
<p>Quantitative Trait Loci (QTL) are regions in the genome that are associated with variation in a quantitative trait. Quantitative traits are phentoypes that can be measured on a continuous scale, like height, weight, etc.</p>
<p>QTL analysis (or QTL mapping) is typically done on experimental populations to find genes which contribute to the heritability of traits. Phenotype and genetic marker data are collected from every individual in the population. The general concept of QTL mapping is that we can then calculate the correlation between genotypes and phenotypes at each marker position and test whether they show a statistically significant association.</p>
<p>Let’s consider the famous example of <span class="citation">Doebley and Stec (<a href="#ref-Doebley285">1991</a>)</span>, who assessed the variation of traits that discriminate commercial maize from its native relative teosinte. Teosinte is much smaller than maize as we know it today and one teosinte plant produces many ears, each of which has only two rows of seeds. But even though maize and teosinte look so completely different, they are still able to produce viable offspring together. <span class="citation">Doebley and Stec (<a href="#ref-Doebley285">1991</a>)</span> utilized this and crossed the two plant species to produce an F1 generation, which were in turn self-pollinated. The resulting F2 population of maize-teosinte-hybrids showed a wide range of intermediate parental morphologies. Each of the F2 offspring was then genotyped at 58 locations in the genome, so that the quantitative trait information on morphology could be correlated with the genetic map. This analysis revealed that most of the morphological variation between maize and teosinte were the result of changes in only a handful of genes, one of which is the <em>tb1</em> (<em>teosinte branched 1</em>) gene.</p>
<div id="recombinant-inbred-lines-rils" class="section level3">
<h3><span class="header-section-number">4.1.1</span> Recombinant Inbred Lines (RILs)</h3>
<p>RILs are experimental sister populations that have been produced by a very specific back-crossing scheme. The process is similar to Doebley and Stec’s crossing of maize and teosinte: two homozygous parents are crossed to produce an F1 generation. Following the laws of genetics, each offspring’s genome consists of a random combination of parental alleles and crossover (or recombination) events. Depending on the design, F1 offspring are usually either selfed or mated with a sibling to introduce another level of genetic recombination. The final generation is then inbred for many generations to obtain a collection of homozygous sister lines, each with a unique mosaic genome of parental alleles <span class="citation">(Pollard <a href="#ref-Pollard2012">2012</a>)</span>.</p>
</div>
<div id="qtl-analysis-in-r" class="section level3">
<h3><span class="header-section-number">4.1.2</span> QTL analysis in R</h3>
<div id="the-qtl-package" class="section level4">
<h4><span class="header-section-number">4.1.2.1</span> The “qtl” package</h4>
<p>The best established R package for QTL mapping is <span class="citation">Broman et al. (<a href="#ref-Broman2003">2003</a>)</span>’s <a href="http://www.rqtl.org/"><strong>qtl</strong> package</a>. It implements several techniques for finding QTLs, like Hidden Markov Models (HMM), interval mapping, Haley-Knott regression and multiple imputation. It is very well documented and comes with extensive example data and code.</p>
<p>Here, I will introduce you to a basic QTL mapping workflow using the examples given in the package documentation and refer you to more complex analysis options where applicable.</p>
<div id="installation-and-loading-the-package" class="section level5 unnumbered">
<h5>Installation and loading the package</h5>
<p>If this is the first time you are using the <strong>qtl</strong> package, you need to install it from CRAN. The following line of code checks whether you already have the package, and if not installs it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pkg =<span class="st"> &quot;qtl&quot;</span>
if (<span class="kw">system.file</span>(<span class="dt">package =</span> pkg) ==<span class="st"> &#39;&#39;</span>) <span class="kw">install.packages</span>(pkg)</code></pre></div>
<p>You can then load the package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(qtl)</code></pre></div>
</div>
<div id="loading-the-data" class="section level5 unnumbered">
<h5>Loading the data</h5>
<p>I will be using the example data on murine hypertension that is provided in the package <span class="citation">(Sugiyama et al. <a href="#ref-Sugiyama200170">2001</a>)</span>. The <strong>summary()</strong> function shows you the main properties of the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(hyper)
<span class="kw">summary</span>(hyper)</code></pre></div>
<pre><code>##     Backcross
## 
##     No. individuals:    250 
## 
##     No. phenotypes:     2 
##     Percent phenotyped: 100 100 
## 
##     No. chromosomes:    20 
##         Autosomes:      1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 
##         X chr:          X 
## 
##     Total markers:      174 
##     No. markers:        22 8 6 20 14 11 7 6 5 5 14 5 5 5 11 6 12 4 4 4 
##     Percent genotyped:  47.7 
##     Genotypes (%):    
##           Autosomes:    BB:50.1  BA:49.9 
##        X chromosome:    BY:53.0  AY:47.0</code></pre>
<p>The <strong>plot()</strong> function produces plots showing missing genotypes, the marker positions and the distribution of phenotypes or traits. This will give you a first idea of your data.</p>
<p>The <a href="http://www.rqtl.org/tutorials/rqtltour.pdf">package manual</a> includes a description of various additional plotting functions, which I won’t cover here. I also advise to examine each object with <strong>head()</strong> or <strong>summary()</strong> after you ran a function to get a feel for your data and the various steps you are taking in the analysis.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(hyper)</code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-4-1.png" width="768" style="display: block; margin: auto;" /></p>
</div>
<div id="genetic-map-estimation" class="section level5 unnumbered">
<h5>Genetic map estimation</h5>
<p>Before we proceed with the analysis, I typically recommend to replace the existing genetic map with an estimated one to reduce the potential errors. The genetic map represents all markers on a chromosome in a linear fashion. The <strong>est.map()</strong> function applies a Hidden Markov Model <span class="citation">(E. S. Lander and Green <a href="#ref-Lander01041987">1987</a>)</span> to estimate the map with an assumed genotyping error rate (<em>error.prob</em>).</p>
<p>Here, we can also specify the mapping function (<em>map.function</em>) that we want to use to convert genetic distance to recombination fraction. The distance between two markers is usually given as a unit of genetic linkage, called <em>centimorgan (cM)</em> One cM represents the distance with an average of 0.01 crossover events in one generation (i.e. 1% recombination). However, this representation of distance underestimates the actual recombination fraction, which is inherently not additive. With increasing distance the chance of double crossovers increases, so that they are in a way “invisible” to the traditional estimation of recombination distance.</p>
<p>Another reason why genetic maps based on recombination are biased is crossover interference, which described the phenomenon that a crossover event reduces the likelihood of another recombination event occur close by.</p>
<p>To correct for such biases, we can choose from the following mapping functions:</p>
<ul>
<li><strong>Haldane’s</strong> is the simplest mapping function and assumes a Poisson distribution for crossover events and does not consider interference.</li>
<li><strong>Kosambi’s</strong> mapping function also considers interference and double crossovers but it can not calculate joint recombination probabilities for more than three loci.</li>
<li><strong>Carter-Falconer’s</strong> mapping function can be extended to more complex interference rates.</li>
<li><strong>Morgan’s</strong> mapping function assumes complete interference.</li>
</ul>
<p>The two most widely used mapping functions are Haldane’s (the default) and Kosambi’s. For this example, using Haldane’s should be sufficient. Because our example is a backcross, we can assume no interference, meaning that all crossovers are independent <span class="citation">(Lynch and Walsh <a href="#ref-lynch1998genetics">1998</a>)</span>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">newmap &lt;-<span class="st"> </span><span class="kw">est.map</span>(hyper, <span class="dt">error.prob =</span> <span class="fl">0.0001</span>, <span class="dt">map.function =</span> <span class="st">&quot;haldane&quot;</span>)
hyper &lt;-<span class="st"> </span><span class="kw">replace.map</span>(hyper, newmap)</code></pre></div>
<p>We can now estimate the recombination fractions between all pairs of markers. The <strong>est.rf()</strong> function also calculates the LOD scores. LOD stands for “likelihood of the odds” and is a measure of linkage. In QTL mapping we calculate LOD scores for the genetic markers and a threshold, above which we consider a QTL statistically significant in its association with the trait.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper &lt;-<span class="st"> </span><span class="kw">est.rf</span>(hyper)</code></pre></div>
<p>The <strong>calc.errorlod()</strong> function calculates the genotyping errors according to <span class="citation">Lincoln and Lander (<a href="#ref-Lincoln1992604">1992</a>)</span>. We can see which markers have an error LOD above a certain threshold (cutoff) with the <strong>top.errorlod()</strong> function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper &lt;-<span class="st"> </span><span class="kw">calc.errorlod</span>(hyper, <span class="dt">error.prob =</span> <span class="fl">0.0001</span>)
te &lt;-<span class="st"> </span><span class="kw">top.errorlod</span>(hyper, <span class="dt">cutoff =</span> <span class="dv">3</span>)
te</code></pre></div>
<pre><code>##   chr  id   marker errorlod
## 1   4 102 D4Mit288 3.324582
## 2   4 107 D4Mit111 3.262205
## 3   4 216 D4Mit214 3.261092
## 4  11  57 D11Mit82 3.021105
## 5  11 118 D11Mit82 3.021105</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper.clean &lt;-<span class="st"> </span>hyper
for(i in <span class="dv">1</span>:<span class="kw">nrow</span>(te)) {
  chr &lt;-<span class="st"> </span>te$chr[i]
  id &lt;-<span class="st"> </span>te$id[i]
  mar &lt;-<span class="st"> </span>te$marker[i]
  hyper.clean$geno[[chr]]$data[hyper$pheno$id ==<span class="st"> </span>id, mar] &lt;-<span class="st"> </span><span class="ot">NA</span>
}</code></pre></div>
</div>
<div id="finding-qtls" class="section level5 unnumbered">
<h5>Finding QTLs</h5>
<p>Now, we can proceed with the central step: mapping the QTLs.</p>
<p>Because the individuals in QTL studies are genotyped at specific marker locations throughout the genome, we inherently have to deal with the missing information about genotypes between markers. Hidden Markov Models (HMM) can help us overcome this problem by calculating genotype probabilities between markers based on the joint genotype distribution.</p>
<p>We first need to calculate these genotype probabilities using the <strong>calc.genoprob()</strong> function. We can define several parameters, like step size, the amount of error we want to allow for, the mapping function and step width. Here, we want to calculate genotype probabilities for every cM (step = 1), with a fixed step width and an error probability of 0.0001. As above, we are again using Haldane’s mapping function.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper.clean &lt;-<span class="st"> </span><span class="kw">calc.genoprob</span>(hyper.clean, <span class="dt">step =</span> <span class="dv">1</span>, <span class="dt">error.prob =</span> <span class="fl">0.0001</span>, <span class="dt">map.function =</span> <span class="st">&quot;haldane&quot;</span>, <span class="dt">stepwidth =</span> <span class="st">&quot;fixed&quot;</span>)</code></pre></div>
<p>The simplest QTL model, we can run is single-QTL marker regression or interval mapping using the <strong>scanone()</strong> function.</p>
<p>These simple methods can give a good estimation of QTLs but they can also introduce bias, especially with multiple QTL in close proximity. More advanced mapping approaches, like Composite Interval Mapping (CIM) are discussed later on.</p>
<p>The first parameter we want to specify is the phenotype(s) and model (e.g. parametric or non-parametric) for mapping. Here, we want to use the first phenotype, i.e. the first column in our phenotype matrix, which follows a normal distribution. We can see the phenotype matrix by calling:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper.clean$pheno</code></pre></div>
<p>Then, we need to specify the mapping algorithm we want to use. We can choose from several options. Here, I will only present the practical implications for each method. For a full discussion of the mathematical principles, see <span class="citation">Lynch and Walsh (<a href="#ref-lynch1998genetics">1998</a>)</span>.</p>
<ul>
<li><strong>Marker regression</strong>: Marker regression is by far the simplest approach to QTL mapping. Here, we calculate the association between phenotype and genotype at each marker position independently.</li>
</ul>
<p>Because it is so simple, marker regression is seldom recommended to use. With interval mapping, a phenotype ~ genotype association analysis is performed for each flanking marker pair independently. This improves the approximation and gives confidence regions around QTL.</p>
<ul>
<li><strong>EM (Expectation-Maximization) algorithm</strong>: EM is usually applied to maximum likelihood (ML) analyses of mixed models. It is an iterative process of calculating conditional probabilities and updating the ML estimates. This process is repeated until the estimates converge <span class="citation">(E. S. Lander and Botstein <a href="#ref-Lander185">1989</a>)</span>. If we have a reasonably dense marker map, the EM algorithm will converge on the global maximum.</li>
<li><strong>(Extended) Haley-Knott regression</strong>: Haley-Knott regression uses a simpler model than the EM algorithm <span class="citation">(Haley and Knott <a href="#ref-Haley1992">1992</a>)</span>. The extended Haley-Knott regression also considers variance is therefore gives improved approximations. Haley-Knott regression can give good approximations of the likelihood profiles for ML interval mapping but with more complex cases, it can be heavily biased.</li>
<li><strong>Multiple imputation</strong>: This method uses multiple rounds of imputing the unknown genotypes between markers and combines them into a final imputation model <span class="citation">(Ś. Sen and Churchill <a href="#ref-Sen371">2001</a>)</span>. This allows us to perform a simple analysis of variance at each position in the genome. Multiple imputation needs much more computational power than simpler methods and will usually not outperform them with single-QTL models (it is much more advantageous with more complex multi-QTL models, however).</li>
</ul>
<p>Here, I will show QTL mapping examples for the EM algorithm and for multiple imputation:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># EM algorithm</span>
out.em &lt;-<span class="st"> </span><span class="kw">scanone</span>(hyper.clean, <span class="dt">pheno.col =</span> <span class="dv">1</span>, <span class="dt">model =</span> <span class="st">&quot;normal&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;em&quot;</span>)</code></pre></div>
<p>The multiple imputation method requires the use of the <strong>sim.geno()</strong> function before we call the mapping function. It calculates the joint genotype distribution based on the available marker information and uses it to perform the imputation of missin genotypes. With the <strong>n.draws</strong> parameter, we define how many imputations will be run. The more imputations steps we run, the more precise the genotypes but with increasing cost of computational time and power.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">hyper.clean &lt;-<span class="st"> </span><span class="kw">sim.geno</span>(hyper.clean, <span class="dt">step =</span> <span class="dv">1</span>, <span class="dt">n.draws =</span> <span class="dv">100</span>, <span class="dt">error.prob =</span> <span class="fl">0.0001</span>, <span class="dt">map.function =</span> <span class="st">&quot;haldane&quot;</span>, <span class="dt">stepwidth =</span> <span class="st">&quot;fixed&quot;</span>)
out.imp &lt;-<span class="st"> </span><span class="kw">scanone</span>(hyper.clean, <span class="dt">pheno.col =</span> <span class="dv">1</span>, <span class="dt">model =</span> <span class="st">&quot;normal&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;imp&quot;</span>)</code></pre></div>
<p>Now that we have a LOD score for each marker position, we want to know which positions are significantly associated with the phenotype. To determine this, we will use the <strong>scanone()</strong> function again, but this time we want to calculate the genome-wide LOD score threshold with a permutation test. Above this threshold we can consider a QTL to be statistically significant. Here, I am using a similar call as before, but I am specifying that we want to use 1000 permutations.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">operm.imp &lt;-<span class="st"> </span><span class="kw">scanone</span>(hyper.clean, <span class="dt">pheno.col =</span> <span class="dv">1</span>, <span class="dt">model =</span> <span class="st">&quot;normal&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;imp&quot;</span>, <span class="dt">n.perm =</span> <span class="dv">1000</span>)</code></pre></div>
<pre><code>## Doing permutation in batch mode ...</code></pre>
<p>The <strong>summary()</strong> function will tell us our genome-wide LOD score threshold for a given significance value (here 0.05).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lod &lt;-<span class="st"> </span><span class="kw">summary</span>(operm.imp, <span class="dt">alpha =</span> <span class="fl">0.05</span>)
lod</code></pre></div>
<pre><code>## LOD thresholds (1000 permutations)
##     lod
## 5% 2.28</code></pre>
<p>And now we can also refine our QTL results by including the significance threshold. This will give us the LOD score and estimated p-values for each marker above the threshold, around which we can now assume to have found a QTL for our trait of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">summary</span>(out.imp, <span class="dt">perms =</span> operm.imp, <span class="dt">alpha =</span> <span class="fl">0.05</span>, <span class="dt">pvalues =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>##           chr   pos  lod pval
## c1.loc97    1 100.3 3.65 0.00
## D4Mit164    4  41.6 8.08 0.00
## D15Mit152  15  37.3 2.34 0.04</code></pre>
<p>Now that we have our QTL, we can plot them with the <strong>plot()</strong> function. Here, I am plotting the results from both, the EM algorithm and the multiple imputation method. As expected, they do not differ much. The horizontal dotted line shows the genome-wide LOD threshold and our two significant QTL pop up nicely on chromosomes 1 and 4.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(out.em, <span class="dt">col =</span> <span class="st">&quot;blue&quot;</span>)
<span class="kw">plot</span>(out.imp, <span class="dt">col =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">add =</span> <span class="ot">TRUE</span>)
<span class="kw">abline</span>(<span class="dt">h =</span> lod[<span class="dv">1</span>], <span class="dt">lty =</span> <span class="dv">2</span>)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="kw">c</span>(<span class="st">&quot;EM algorithm&quot;</span>,<span class="st">&quot;Mult. imputation&quot;</span>), <span class="dt">col =</span> <span class="kw">c</span>(<span class="st">&quot;blue&quot;</span>, <span class="st">&quot;red&quot;</span>), <span class="dt">lty =</span> <span class="dv">1</span>, <span class="dt">lwd =</span> <span class="dv">3</span>)</code></pre></div>
<p><img src="_main_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
</div>
<div id="qtl-interaction-mapping" class="section level5 unnumbered">
<h5>QTL interaction mapping</h5>
</div>
<div id="covariates-in-qtl-models" class="section level5 unnumbered">
<h5>Covariates in QTL models</h5>
</div>
</div>
<div id="qtl-analysis-using-bayesian-interval-mapping-qtlbim-package" class="section level4">
<h4><span class="header-section-number">4.1.2.2</span> QTL Analysis using Bayesian Interval Mapping (“qtlbim” package)</h4>
</div>
</div>
</div>
<div id="gene-x-environment-interactions" class="section level2">
<h2><span class="header-section-number">4.2</span> Gene x Environment interactions</h2>
</div>
<div id="variance-and-heritability" class="section level2">
<h2><span class="header-section-number">4.3</span> Variance and heritability</h2>

</div>
</div>
<h3> Placeholder</h3>
<div id="refs" class="references">
<div id="ref-Doebley285">
<p>Doebley, J, and A Stec. 1991. “Genetic Analysis of the Morphological Differences Between Maize and Teosinte.” <em>Genetics</em> 129 (1). Genetics: 285–95. <a href="http://www.genetics.org/content/129/1/285" class="uri">http://www.genetics.org/content/129/1/285</a>.</p>
</div>
<div id="ref-Pollard2012">
<p>Pollard, Daniel A. 2012. “Design and Construction of Recombinant Inbred Lines.” In <em>Quantitative Trait Loci (Qtl): Methods and Protocols</em>, edited by Scott A. Rifkin, 31–39. Totowa, NJ: Humana Press. doi:<a href="https://doi.org/10.1007/978-1-61779-785-9_3">10.1007/978-1-61779-785-9_3</a>.</p>
</div>
<div id="ref-Broman2003">
<p>Broman, Karl W., Hao Wu, Saunak Sen, and Gary A. Churchill. 2003. “R/Qtl: QTL Mapping in Experimental Crosses.” <em>Bioinformatics</em> 19: 889–90.</p>
</div>
<div id="ref-Sugiyama200170">
<p>Sugiyama, Fumihiro, Gary A. Churchill, David C. Higgins, Conrado Johns, Konstatinos P. Makaritsis, Haralambos Gavras, and Beverly Paigen. 2001. “Concordance of Murine Quantitative Trait Loci for Salt-Induced Hypertension with Rat and Human Loci.” <em>Genomics</em> 71 (1): 70–77. doi:<a href="https://doi.org/http://dx.doi.org/10.1006/geno.2000.6401">http://dx.doi.org/10.1006/geno.2000.6401</a>.</p>
</div>
<div id="ref-Lander01041987">
<p>Lander, E S, and P Green. 1987. “Construction of Multilocus Genetic Linkage Maps in Humans.” <em>Proceedings of the National Academy of Sciences</em> 84 (8): 2363–7. <a href="http://www.pnas.org/content/84/8/2363.abstract" class="uri">http://www.pnas.org/content/84/8/2363.abstract</a>.</p>
</div>
<div id="ref-lynch1998genetics">
<p>Lynch, M., and B. Walsh. 1998. <em>Genetics and Analysis of Quantitative Traits</em>. Sinauer. <a href="https://books.google.de/books?id=UhCCQgAACAAJ" class="uri">https://books.google.de/books?id=UhCCQgAACAAJ</a>.</p>
</div>
<div id="ref-Lincoln1992604">
<p>Lincoln, Stephen E., and Eric S. Lander. 1992. “Systematic Detection of Errors in Genetic Linkage Data.” <em>Genomics</em> 14 (3): 604–10. doi:<a href="https://doi.org/http://dx.doi.org/10.1016/S0888-7543(05)80158-2">http://dx.doi.org/10.1016/S0888-7543(05)80158-2</a>.</p>
</div>
<div id="ref-Lander185">
<p>Lander, E S, and D Botstein. 1989. “Mapping Mendelian Factors Underlying Quantitative Traits Using Rflp Linkage Maps.” <em>Genetics</em> 121 (1). Genetics: 185–99. <a href="http://www.genetics.org/content/121/1/185" class="uri">http://www.genetics.org/content/121/1/185</a>.</p>
</div>
<div id="ref-Haley1992">
<p>Haley, C S, and S A Knott. 1992. “A Simple Regression Method for Mapping Quantitative Trait Loci in Line Crosses Using Flanking Markers.” <em>Heredity</em> 69 (4). The Genetical Society of Great Britain: 315–24. doi:<a href="https://doi.org/http://dx.doi.org/10.1038/hdy.1992.131">http://dx.doi.org/10.1038/hdy.1992.131</a>.</p>
</div>
<div id="ref-Sen371">
<p>Sen, Śaunak, and Gary A. Churchill. 2001. “A Statistical Framework for Quantitative Trait Mapping.” <em>Genetics</em> 159 (1). Genetics: 371–87. <a href="http://www.genetics.org/content/159/1/371" class="uri">http://www.genetics.org/content/159/1/371</a>.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="population-genetics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="genetics-of-complex-diseases.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"download": ["_main.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
